README.md

```diff
@@ -285,6 +285,108 @@ There are many places in the format for compatible extensions:
+| `PAR` | Reserved for the future when an extension replaces `PAR1` |
+| `PER` | Reserved for the future when an extension replaces `PARE` |
+| `ASF` | Apache |
+| `AWS` | Amazon |
+| `CDH` | Cloudera |
+| `CRM` | Salesforce |
+| `DBR` | Databricks |
+| `EXP` | Apache/Experimental |
+
+To reserve additional magic numbers, file a JIRA and send a PR.
+
+The magic is 3 bytes because it is always followed by the 0 byte, the thrift
+field stop byte. Together this defines a 4 byte magic number, which can be used
+in place of existing parquet magic numbers.
+
+#### An example FileMetaData replacement and migration plan
+
+Consider the case of extending `FileMetaData` a full replacement. That is
+the new encoding contains all the information of `FileMetaData` and readers
+that know about it can elide parsing the current thrift `FileMetaData`. This
+extension has additional considerations and requirements.
+
+First, observe that `FileMetaData` is located between the last column chunk
+and the 4-byte length plus 4-byte `PAR1` magic. Sophisticated parquet readers,
+typically read the tail of files speculatively and expect to find the full
+footer in that fetch. Thus our extension must be decodable from the end of the
+file, and without having to fetch the full old `FileMetaData` thrift encoding.
+As a corollary finding the bounds of the extension should not require thrift
+parsing.
+
+To satisfy these requirements we define our `FileMetaData` extension as:
+
+    N bytes: the new `FileMetaData` replacement - in some encoding
+    4 bytes: little endian crc32 of the previous N bytes
+    4 bytes: N in little endian
+    4 bytes: little endian crc32 of N
+    3 bytes: 3-byte magic extension from the table above
+
+Each field plays its role to satisfy the requirements. In reverse order:
+1. 3-byte magic extension: as per the specification. When this new
+`FileMetaData` replaces the old, we can replace the thrift `FileMetaData`
+including the trailing 8 bytes and replace them with our extension plus the
+null byte verbatim.
+2. `le32(N)` + `crc32(N)`: The pair of len and its crc32 is useful to validate
+that the length is correct. Otherwise we might be tripped to read an
+unspecified number of bytes only to later find their crc32 does not match.
+3. `crc32(bytes)`: The crc32 of the new `FileMetaData` itself is important to
+avoid reading corrupt or erroneous metadata.
+4. The bytes of the encoding. This should be in our new encoding, for the sake
+of argument flatbuffers, prefixed with an ID so that we as we experiment can
+distinguish different versions of metadata.
+
+The development and migration plan might look like:
+1. A period where the new `FileMetaData` will be written after the old, with a
+non-reserved 3 byte magic, say `DBR`.
+2. Once the format stabilizes and is considered final, it is brought to the
+parquet commitee for ratification.
+3. When ratified the extension is moved to an approved state and takes the
+reserved 3 byte magic `PAR`.
```

Comment:
If this is intended to be a temporary state while the new PAR footer format is being iterated on, maybe we don't need to reserve names like this as "PAR" will be the eventual replacement excluding other names.

It seems to me that this proposal would be better described as a path toward evolving the parquet footer while preserving compatibility for a while. The new footer replacing the existing one eventually.

I think extensibility would be useful as well but those two are not the same thing. By extensibility, I would expect users to have a safe, efficient way to add metadata to the file.

Reply:
My main concern with this type of proposal, is the "considered final" piece.  I think if there is closed development here, some extent the incentives of getting to a new 'PAR' are potentially misaligned.  

I like how we are currently iterating in the open on the proposals so everyone has a chance to validate them and provide input.  From a community perspective there is obviously nothing that can be done to stop vendors from having custom metadata in the long run, but at least it would be nice to try to provide guidance for the interim process to encourage openness (maybe the table  above could also contain an optional link to a public repo/gist so that the community can keep tabs on the different approaches?)

Reply:
This was discussed in Parquet Sync and replaced with a UUID.